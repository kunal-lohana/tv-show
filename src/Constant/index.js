export const Constants = {
  NAV_Gener: "Genre",
  CAST: "Cast",
  EPISODES: "Episodes",
  SHOW_DETAILS_BTN: "Show Details",
  OFFICIAL_SITE: "Offical Site",
  DESCRIPTION: "Description",
  SCHEDULE_AVAILABILITY: "Schedule",
  PAGE_NOT_FOUND: "PAGE NOT FOUND",
  ERROR_CODE: "404",
  HOME_PAGE: "Home"
};
