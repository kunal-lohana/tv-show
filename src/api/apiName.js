export const TV_SHOW_List = "getTvShowList";
export const TV_SHOW_DETAIL = "getTvShowDetail";
export const SHOW_CAST_DETAIL = "getCastDetail";
export const EPISODES_DETAIL = "getEpisodesDetail";
export const SEARCH_SHOW = "searchShow";
