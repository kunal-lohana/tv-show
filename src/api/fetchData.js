import axios from "axios";
import config from "../config/config";

export const fetchData = ({ apiName, params }) => {
  const requests = {
      get: url => axios.get(`${config.API_URL}${url}`),
      post: (url, body) => axios.post(`${config.API_URL}${url}`, body)
    },
    Api = {
      getTvShowList: params => requests.get(`shows`, params),
      getTvShowDetail: params =>
        requests.get(
          `shows/${params.id}?embed[]=episodes&embed[]=cast`,
          params
        ),
      searchShow: params =>
        requests.get(`/search/shows?q=${params.query}`, params)
    };
  return Api[apiName](params);
};
