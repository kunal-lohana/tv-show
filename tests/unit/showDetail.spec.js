import { shallowMount, mount, config, createLocalVue } from "@vue/test-utils";
import { showDescription, validateCaste, castDetails } from "./testingData";
import Vuetify from "vuetify";
import Vue from "vue";
import VueRouter from "vue-router";
import ShowDetailComponent from "@/components/ShowDetail/ShowDetail.vue";
import { Constants } from "@/Constant/index";
import { fetchData } from "@/api/fetchData.js";
import { routes } from "@/router/index.js";
import { TV_SHOW_DETAIL } from "@/api/apiName.js";

config.showDeprecationWarnings = false;
Vue.config.slient = false;

jest.mock("@/api/fetchData.js", () => ({
  fetchData: jest.fn(async () => {
    const showDescription = await require("./testingData").showDescription;
    return Promise.resolve({ data: showDescription });
  })
}));

describe("Show Detail Component ", () => {
  let wrapper;
  let localVue = createLocalVue();
  localVue.use(VueRouter);
  Vue.use(Vuetify);
  let vuetify = new Vuetify();

  const router = new VueRouter({
    routes
  });
  beforeEach(() => {
    wrapper = mount(ShowDetailComponent, {
      router,
      localVue,
      vuetify,
      data() {
        return {
          showDescription: showDescription,
          constant: Constants,
          loading: false
        };
      }
    });
  });
  it("should match snapshot", async () => {
    const wrapper = shallowMount(ShowDetailComponent, {
      router,
      localVue,
      vuetify
    });
    await wrapper.setData({
      showDescription: showDescription,
      constant: Constants,
      loading: false,
      castDetails: castDetails,
      episodeDetails: showDescription._embedded.episodes
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("check show name", async () => {
    const wrapper = shallowMount(ShowDetailComponent, {
      router,
      localVue,
      vuetify,
      data() {
        return {
          showDescription: showDescription
        };
      }
    });
    await wrapper.setData({
      loading: false
    });
    const showName = wrapper.find('[data-test="show-name"]');
    expect(showName.text()).toEqual(showDescription.name);
  });
  it("check rating", async () => {
    await wrapper.setData({
      loading: false
    });
    const rating = wrapper.find('[data-test="show-rating"]');
    expect(rating.text()).toEqual(`${showDescription.rating.average / 2} (5)`);
  });

  it("check genre", async () => {
    await wrapper.setData({
      loading: false
    });
    const genre = wrapper.find('[data-test="show-genre"]');
    expect(genre.text()).toEqual(showDescription.genres.join(","));
  });

  it("check language", async () => {
    await wrapper.setData({
      loading: false
    });
    const language = wrapper.find('[data-test="show-language"]');
    expect(language.text()).toEqual(showDescription.language);
  });

  it("check summary", async () => {
    await wrapper.setData({
      loading: false
    });
    const summary = wrapper.find('[data-test="show-summary"] > p');
    expect(summary.html()).toEqual(showDescription.summary);
  });

  it("check time", async () => {
    await wrapper.setData({
      loading: false
    });
    const time = wrapper.find('[data-test="show-time"]');
    expect(time.text()).toEqual(showDescription.schedule.time);
  });

  it("check days", async () => {
    await wrapper.setData({
      loading: false
    });
    const days = wrapper.find('[data-test="show-days"]');
    expect(days.text()).toEqual(showDescription.schedule.days.join(","));
  });

  it("check constant of offical site", async () => {
    await wrapper.setData({
      loading: false
    });
    const constantSite = wrapper.find('[data-test="constant-site"]');
    expect(constantSite.text()).toEqual(wrapper.vm.constant.OFFICIAL_SITE);
  });

  it("check constant of schedule", async () => {
    await wrapper.setData({
      loading: false
    });
    const constantSchedule = wrapper.find('[data-test="constant-schedule"]');
    expect(constantSchedule.text()).toEqual(
      wrapper.vm.constant.SCHEDULE_AVAILABILITY
    );
  });

  it("check constant of description", async () => {
    await wrapper.setData({
      loading: false
    });
    const constantDescription = wrapper.find(
      '[data-test="constant-description"]'
    );
    expect(constantDescription.text()).toEqual(wrapper.vm.constant.DESCRIPTION);
  });

  it("check constant of episodes", async () => {
    const constantEpisodes = wrapper.find('[data-test="constant-episodes"]');
    expect(constantEpisodes.text()).toEqual(wrapper.vm.constant.EPISODES);
  });

  it("check constant of cast", async () => {
    const constantCast = wrapper.find('[data-test="constant-cast"]');
    expect(constantCast.text()).toEqual(wrapper.vm.constant.CAST);
  });

  it("Created lifecycle hook called getShowDetails method", async () => {
    const getShowDetails = jest.spyOn(
      ShowDetailComponent.methods,
      "getShowDetails"
    );
    mount(ShowDetailComponent, {
      router,
      localVue,
      vuetify,
      data() {
        return {
          showDescription: showDescription
        };
      }
    });
    expect(getShowDetails).toHaveBeenCalled();
  });

  it("getShowDetails method execute fetchData", async () => {
    await fetchData.mockImplementationOnce(async () => {
      const showDescription = await require("./testingData").showDescription;
      return Promise.resolve({ data: showDescription });
    });
    const wrapper = mount(ShowDetailComponent, {
      router,
      localVue,
      vuetify
    });
    await wrapper.vm.getShowDetails(2);
    expect(fetchData).toHaveBeenCalled();
  });

  it("call fetchData will set showDescription, castDetail", async () => {
    await fetchData.mockImplementationOnce(async () => {
      const showDescription = await require("./testingData").showDescription;
      return Promise.resolve({ data: showDescription });
    });
    const wrapper = mount(ShowDetailComponent, {
      router,
      localVue,
      vuetify
    });
    const result = await fetchData({
      apiName: TV_SHOW_DETAIL,
      params: { id: 2 }
    });
    expect(fetchData).toHaveBeenCalled();
    expect(wrapper.vm.showDescription).toEqual(result.data);
  });

  it("getShowDetails method call and fetchData return blank array, navigate to pagenotfound", async () => {
    const spyMethod = jest.spyOn(ShowDetailComponent.methods, "getShowDetails");
    const wrapper = mount(ShowDetailComponent, {
      router,
      localVue,
      vuetify
    });
    await wrapper.setData({
      loading: false,
      showDescription: showDescription
    });
    await wrapper.vm.getShowDetails();
    expect(spyMethod).toHaveBeenCalled();
    expect(wrapper.vm.loading).toBeFalsy();
  });

  it("validateCasteDetails method unique cast object", async () => {
    expect(validateCaste.length).toBe(2);
    const result = await wrapper.vm.validateCasteDetails(validateCaste);
    expect(result.length).toBe(1);
  });
});
