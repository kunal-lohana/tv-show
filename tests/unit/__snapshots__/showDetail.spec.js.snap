// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Show Detail Component  should match snapshot 1`] = `
<v-card-stub loaderheight="4" tag="div" color="secondary" dark="true" app="" class="mx-auto my-12">
  <div>
    <v-row-stub tag="div" align="center" justify="center">
      <v-col-stub cols="12" md="4" tag="div" align="center">
        <v-img-stub height="300" options="[object Object]" position="center center" src="http://static.tvmaze.com/uploads/images/medium_portrait/1/4185.jpg" transition="fade-transition" data-test="show-image" class="d-flex d-md-none"></v-img-stub>
        <v-img-stub height="300" width="300" options="[object Object]" position="center center" src="http://static.tvmaze.com/uploads/images/medium_portrait/1/4185.jpg" transition="fade-transition" class="d-none d-md-flex"></v-img-stub>
      </v-col-stub>
      <v-col-stub cols="12" md="8" tag="div" align="center">
        <v-card-text-stub>
          <v-card-title-stub datatest="show-name" data-test="show-name" class=" justify-center">Cops</v-card-title-stub>
          <v-rating-stub color="amber" opendelay="0" closedelay="0" ripple="true" backgroundcolor="accent" dense="true" emptyicon="$ratingEmpty" fullicon="$ratingFull" halficon="$ratingHalf" halfincrements="true" length="5" readonly="true" size="14" value="3.9"></v-rating-stub>
          <div class="grey--text ml-4"><span data-test="show-rating">3.9 (5)</span></div>
          <div data-test="show-genre" class="my-2 subtitle-1">
            Action,Crime
          </div>
          <div data-test="show-language" class="my-2 subtitle-1">
            English
          </div>
        </v-card-text-stub>
      </v-col-stub>
    </v-row-stub>
    <v-row-stub tag="div">
      <v-col-stub cols="12" tag="div">
        <v-divider-stub class="mx-4"></v-divider-stub>
        <v-card-text-stub>
          <v-card-title-stub datatest="constant-description" data-test="constant-description" class="pl-0">Description</v-card-title-stub> <span data-test="show-summary"><p><b>COPS</b> follows police officers, constables, and sheriff's deputies during patrols and various police activities by embedding camera crews with their units. The show's formula adheres to a classic <i>cinéma vérité</i> ethos. With no narration or scripted dialog, it depends entirely on the commentary of the officers and on the actions of the people with whom they come into contact.</p></span>
        </v-card-text-stub>
      </v-col-stub>
    </v-row-stub>
    <v-divider-stub class="mx-4"></v-divider-stub>
    <v-row-stub tag="div">
      <v-col-stub cols="6" tag="div">
        <v-card-title-stub datatest="constant-schedule" data-test="constant-schedule" class="d-inline-block text-truncate">Schedule
        </v-card-title-stub>
        <v-breadcrumbs-stub divider="-" items="">
          <v-breadcrumbs-item-stub activeclass="v-breadcrumbs__item--disabled" data-test="show-time" items="22:30">
            22:30
          </v-breadcrumbs-item-stub>
          <v-breadcrumbs-item-stub activeclass="v-breadcrumbs__item--disabled" data-test="show-days" items="Monday">
            Monday
          </v-breadcrumbs-item-stub>
        </v-breadcrumbs-stub>
      </v-col-stub>
      <v-divider-stub inset="true" vertical="true" class="my-4"></v-divider-stub>
      <v-col-stub cols="5" tag="div">
        <v-card-title-stub datatest="constant-site" data-test="constant-site">
          Offical Site
        </v-card-title-stub>
        <v-card-actions-stub class="pa-4 "><a d-inline="" target="_blank" href="http://www.cops.com" class="pl-2">
            Cops
          </a></v-card-actions-stub>
      </v-col-stub>
    </v-row-stub>
  </div>
  <v-row-stub tag="div">
    <v-col-stub cols="12" tag="div">
      <v-divider-stub class="mx-4"></v-divider-stub>
      <v-card-title-stub datatest="constant-cast" data-test="constant-cast">
        Cast
      </v-card-title-stub>
      <div>
        <v-container-stub tag="div" grid-list-xs="">
          <v-layout-stub tag="div" row="" wrap="">
            <show-cast-detail-stub data="[object Object]"></show-cast-detail-stub>
          </v-layout-stub>
        </v-container-stub>
      </div>
    </v-col-stub>
  </v-row-stub>
  <v-row-stub tag="div">
    <v-col-stub cols="12" tag="div">
      <v-divider-stub class="mx-4"></v-divider-stub>
      <v-card-title-stub datatest="constant-episodes" data-test="constant-episodes">
        Episodes
      </v-card-title-stub>
      <div>
        <v-container-stub tag="div" grid-list-xs="">
          <v-layout-stub tag="div" row="" wrap="">
            <show-cast-detail-stub data="[object Object]"></show-cast-detail-stub>
            <show-cast-detail-stub data="[object Object]"></show-cast-detail-stub>
          </v-layout-stub>
        </v-container-stub>
      </div>
    </v-col-stub>
  </v-row-stub>
</v-card-stub>
`;
